### PR TITLE
Hopefully reduce occupancy.

### PR DESCRIPTION
**Pull request by fasaxc**
@plwhite this implements the optimization that we discussed.  Would be good if you or Alex could give it a spin in the simulated scale test to see if it actually helps.

* Replace endpoint dict with optimized struct in IpsetManager.
* Only store endpoints that contribute.

### CODE REVIEW

File: calico/felix/ipsets.py [hunk 1]
```diff
@@ -50,8 +50,7 @@ def __init__(self, ip_type):
         # State.
         # Tag IDs indexed by profile IDs
         self.tags_by_prof_id = {}
-        # Endpoint dicts indexed by EndpointId objects
-        self.endpoints_by_ep_id = {}
+        self.endpoint_data_by_ep_id = {}
```
Comment (plwhite) [inline calico/felix/ipsets.py]: You lost a comment here in the change!

File: calico/felix/ipsets.py [hunk 2]
```diff
@@ -361,7 +391,7 @@ def _add_profile_index(self, prof_ids, endpoint_id):
         """
         Notes in the index that an endpoint uses the given profiles.
 
-        :param list[str] prof_ids: List of profile IDs
+        :param set[str] prof_ids: List of profile IDs
```
Comment (plwhite) [inline calico/felix/ipsets.py]: Comment says both set and List. One of these is wrong.

File: calico/felix/ipsets.py [hunk 3]
```diff
@@ -372,7 +402,7 @@ def _remove_profile_index(self, prof_ids, endpoint_id):
         Notes in the index that an endpoint no longer uses any of the
         given profiles.
 
-        :param list[str] prof_ids: List of profile IDs
+        :param set[str] prof_ids: List of profile IDs
```
Comment (plwhite) [inline calico/felix/ipsets.py]: Again, comment inconsistent.

**Comment by plwhite** [comment]
LGTM with a couple of minor comments about comments.

**Comment by plwhite** [comment]
In my initial testing however, this fix made absolutely no difference whatsoever, to the point where I am wondering if the data is kept referenced somewhere else so that we cannot remove it from memory? Just can't explain why it doesn't seem to work.

To add to the sheer entertainment value I put a return at the top of "on_endpoint_update" in ipsets which also had no effect at all, which backs the idea above though I really cannot find any references to objects after being passed in.

**Comment by alexwlchan** [comment]
@plwhite I have literally just emailed you to say that it made no difference in my testing as well. I was feeling a bit dubious, but maybe something else is going on.

**Comment by fasaxc** [comment]
@alexwlchan @plwhite Yay for testing!

**Comment by plwhite** [comment]
If we hadn't tested we'd be finished by now. Just sayin'.

**Comment by fasaxc** [comment]
Pushed my markups.  I think we agreed that it was a good change, even if it didn't affect occupancy.

**Comment by plwhite** [comment]
It turns out that the failure to reduce occupancy is a test artefact. When under heavy enough test load, Felix was just getting behind and so failing to process messages in a timely manner. The impact of this fix is that when it does process the messages it throws away unnecessary data.

Key headline is that in my scale test, creating 10s of thousands of small endpoints, this cut the memory use per endpoint from about 2.5KiB to about 1KiB (resident set - slightly smaller impact on virtual size). That means this is a good fix and does have benefit.

It also means that we should consider further improvements, namely throwing away the unneeded data earlier (we put data on queues that will later be thrown away - better to throw it away before putting on queues). @fasaxc - you should be aware of this.

Merge blocked pending release, then we'll merge this one.

**Comment by matthewdupre** [comment]
Please add an entry to CHANGES.md for this fix.

**Comment by fasaxc** [comment]
@matthewdupre done.

**Pull request merged**